### ISSUE TITLE
[Gitalk:/hardware/en/maixsense/maixsense-a010/maixsense-a010.html] MaixSense-A010 - Sipeed Wiki

### ISSUE THREAD

**Issue by wonderfullook** [sipeed/sipeed_wiki#400]
https://wiki.sipeed.com/hardware/en/maixsense/maixsense-a010/maixsense-a010.html

**Comment by jake653** [comment]
What is the signal level on the UART connection, is it 3.3V logic?
Can it be powered from the 5V pin with no USB connection? a current spec would be useful.

**Comment by wonderfullook** [comment]
> What is the signal level on the UART connection, is it 3.3V logic? Can it be powered from the 5V pin with no USB connection? a current requirement spec would be useful.

Yes, UART connection is 3.3V logic, and boot from the 5V pin is ok.

![image](https://user-images.githubusercontent.com/38576701/227478666-6ec6171a-4f6a-4a8f-9865-be710398b3f1.png)

**Comment by jake653** [comment]
#2 unclear references in the documentation:
"Button function:  "Next" on the left and "Switch" on the right."
 Buttons seem to have low processing priority, unsure what can be set.
I did randomly get UNIT to change to 1, which narrowed the range.

"ANTIMMI:Automatic anti-multi-machine interference AT setting."
 What is that, I find no references in Google (translation issue)

I am unable to find any projects or reviews written by users online. Which is usually how I find answers.

**Comment by rarestohanean** [comment]
On Windows connected with USB-C, I can see two COM ports.
I receive data from those ports at the same time? How? Even and odd frames?
Thank you!

**Comment by shinde-ak** [comment]
can we flash the modified firmware to a010 with post processing ? if so is there any documentation or we need K210 ?

**Comment by sambhavjain98** [comment]
Amazing module, we are impressed with the features and suits exactly our purpose. 
Is the firmware for A010 available? 
We would like to do some modifications to firmware, if its available and we are exploring options to build and flash the A010 directly.

**Comment by jacob-02** [comment]
Is there a possibility to use this with an Arduino board? I have downloaded the MaixSense development board on the Arduino IDE but I don't know which of the board available will be compatible with the MaxiSense-A010.

**Comment by 3space123** [comment]
@jacob-02 Did you ever figure out how to connect the A010 with Arduino?

**Comment by jconnell11** [comment]
Windows Quick Start (from frustrated user):

1. Get the big (slow) 45MB download, not just the update of comtool from Github. The big download contains a needed visualization plug-in.

2. Use a different USB-C cable. The one I received seemed very flakey and required constant wiggling to keep connected. When you plug the device into a USB port it should stay on and show a live colored image on its back.

3. Start up comtool.exe from the big download (no need to run as admin). 
  a. Make sure Graph mode is selected (use "+" if needed)
  b. Double click "MetaSenseLite" if no black display area
  c. Stretch dialog box vertically until you can see the 0->1 color bar
  d. Find lower numbered of two ports associated with "FTDI"
  e. Default baudrate 115200 is fine (irrelevant?)

4. Hit "Open" which should result in "Connected" and a live image. If you do not see something similar to the back of the device, toggle the "USB" box off and on.

5. Hit the long blue bar to set the capture directory. After this, clicking "Capture" will save a image named <timecode>_unitN.bmp here.

**Comment by LeonelCamposM** [comment]
> Is there a possibility to use this with an Arduino board? I have downloaded the MaixSense development board on the Arduino IDE but I don't know which of the board available will be compatible with the MaxiSense-A010. What is the protocol for using this along with an Arduino board through the UART communication?

> @jacob-02 Did you ever figure out how to connect the A010 with Arduino?

I successfully connected the A010 camera to my Xiao ESP32-S3. My approach was to create a simple bridge between the A010 and my Raspberry Pi 4B using the ESP32. The ESP32 reads the raw UART output from the A010 and redirects it over its serial port to the Raspberry Pi. This setup worked well for me, especially since my Raspberry Pi had issues reading from the FTDI serial converter on the A010 board but could easily read the data via the ESP32’s USB interface.

Here’s the code I used for the setup:

```cpp
Copy code
#include <HardwareSerial.h>

// Define UART parameters
#define UART1_TX_PIN D5
#define UART1_RX_PIN D8
#define BAUD_RATE_1 115200
#define BAUD_RATE_2 921600

HardwareSerial mySerial1(1);  // Use UART1 for device communication

void setup() {
  // Initialize UART0 for Serial Monitor (default)
  Serial.begin(BAUD_RATE_2);
  mySerial1.begin(BAUD_RATE_1, SERIAL_8N1, UART1_RX_PIN, UART1_TX_PIN);

  // Optional: Send initial commands to the camera
  mySerial1.print("AT+FPS=19\r");  // Set frames per second
  delay(100);
  mySerial1.print("AT+DISP=4\r");  // Set display mode
  delay(100);
}

void loop() {
  // Pass data from the camera to the serial output
  while (mySerial1.available()) {
    Serial.write(mySerial1.read());  // Forward the camera's data to the serial monitor
  }
}
```

This code initializes a UART connection between the A010 camera and the ESP32, then forwards the camera's data to the serial monitor. If you're using a different Arduino board, you might need to adjust the TX and RX pin definitions to match your board's configuration.

This approach also adds an overhead to the processing speed, but it is the only solution I find to be able to use the device avoiding the FTDI usb driver issues, hope it helps.

**Comment by des4ire** [comment]
Is it possible to download raw histogram data from this device?

**Comment by jconnell11** [comment]
> Is it possible to download raw histogram data from this device?

AFAIK, it only streams raw pixel data. It is pretty easy to build your own histogram from these values.

**Comment by UnitySe3D** [comment]
Does anyone know how to communicate with it over USB 2.0 using python?

**Comment by jconnell11** [comment]
The top level of [tof_cam](https://github.com/jconnell11/tof_cam) is in Python.

**Comment by UnitySe3D** [comment]
Do you know if there is an alternative that works on windows?

**Comment by UnitySe3D** [comment]
Thanks!

**Comment by gravity0717** [comment]
Hello, I’d like to inquire about a possible frequency dropping issue with the Maixsense A010 sensor.

Question 1: ROS Topic Frequency Fluctuation Issue
While using the Maixsense A010 device, I observed variations in the frequency of ROS (Robot Operating System) topics.

When the device is connected directly: The frequency is about 8 Hz.

[Direct to PC]
average rate: 7.742
min: 0.040s   max: 6.969s   std dev: 0.07520s


When the device is connected via a USB hub: The frequency is about 15 Hz.

[Hub]
average rate: 14.663
min: 0.040s   max: 0.102s   std dev: 0.00996s

Question: What causes this frequency fluctuation? Additionally, if there are any solutions or configuration methods to ensure a stable frequency, please let me know.

Question 2: USB 1.1 Connection Intent
I have confirmed that the Maixsense A010 device is recognized in USB 1.1 mode when connected to the computer.

Question: Was there a specific reason for intentionally connecting in USB 1.1 mode? In particular, in relation to the frequency fluctuation issue when connected through a USB hub, how does USB 1.1 connection affect it?
